In [30]:
import pandas as pd
import numpy as np

In [18]:
df = pd.read_csv("https://raw.githubusercontent.com/jakevdp/data-CDCbirths/master/births.csv")
print('Dataframe shape:', df.shape)


Dataframe shape: (15547, 5)


In [19]:
print(df.columns.tolist())

['year', 'month', 'day', 'gender', 'births']


In [20]:
df.gender.value_counts()

F    7776
M    7771
Name: gender, dtype: int64

In [21]:
df.head() #first five observations

year  month  day gender  births
0  1969      1  1.0      F    4046
1  1969      1  1.0      M    4440
2  1969      1  2.0      F    4454
3  1969      1  2.0      M    4548
4  1969      1  3.0      F    4548

In [22]:
df.tail() #last 5 observations

year  month  day gender  births
15542  2008     10  NaN      M  183219
15543  2008     11  NaN      F  158939
15544  2008     11  NaN      M  165468
15545  2008     12  NaN      F  173215
15546  2008     12  NaN      M  181235

We see that that the data is relatively simple–it contains the number of births grouped by date and gender. The tail values show some missing values. Let's investigate further.

In [23]:
df.describe()

year         month           day         births
count  15547.000000  15547.000000  15067.000000   15547.000000
mean    1979.037435      6.515919     17.769894    9762.293561
std        6.728340      3.449632     15.284034   28552.465810
min     1969.000000      1.000000      1.000000       1.000000
25%     1974.000000      4.000000      8.000000    4358.000000
50%     1979.000000      7.000000     16.000000    4814.000000
75%     1984.000000     10.000000     24.000000    5289.500000
max     2008.000000     12.000000     99.000000  199622.000000

In [24]:
df.isnull().sum()

year        0
month       0
day       480
gender      0
births      0
dtype: int64

There are 480 missing values for the day of birth. Also note an error in the max of day (currently 99) which should not be more than 31 depending on the month of the year. The std on day column is also relatively high due to anomalously high values.

In [26]:
months_with_errors = df[df["day"]>31] #months with more than 31 days. 

months_with_errors

year  month   day gender  births
62     1969      1  99.0      F      26
63     1969      1  99.0      M      38
126    1969      2  99.0      F      42
127    1969      2  99.0      M      48
190    1969      3  99.0      F      64
...     ...    ...   ...    ...     ...
14572  1988      4  99.0      F       1
14635  1988      5  99.0      F       1
14696  1988      6  99.0      F       1
14697  1988      6  99.0      M       1
14944  1988     10  99.0      M       1

[350 rows x 5 columns]

In [27]:
df.drop(months_with_errors.index, inplace = True) #Eliminate those 350 rows. The number of rows should be 15547 less 350
df

year  month  day gender  births
0      1969      1  1.0      F    4046
1      1969      1  1.0      M    4440
2      1969      1  2.0      F    4454
3      1969      1  2.0      M    4548
4      1969      1  3.0      F    4548
...     ...    ...  ...    ...     ...
15542  2008     10  NaN      M  183219
15543  2008     11  NaN      F  158939
15544  2008     11  NaN      M  165468
15545  2008     12  NaN      F  173215
15546  2008     12  NaN      M  181235

[15197 rows x 5 columns]

In [28]:
15547-15197 #Confirm if 350

350

In [32]:
list_of_day = df['day'].tolist() #prefer changing to list
day_np=np.array(list_of_day) #modify to np array
#day_np[day_np > 31] = np.nan #find months that indicate day above 31 and replace with nan
#check if months 1,2,3,5,7,8,10,12 have 31 days or less
#check if months 9,4,6 and 11 have days equal or less than 30
day_mean=np.nanmean(day_np)
day_mean
#df['day'].replace(df[df["day"]>31], np.NaN)

15.838078412719984

In [33]:
mean_day = df['day'].mean()
mean_day

15.838078412719984

In [34]:
df.isnull().sum()

year        0
month       0
day       480
gender      0
births      0
dtype: int64

In [36]:
df.describe()

year         month           day         births
count  15197.000000  15197.000000  14717.000000   15197.000000
mean    1979.095940      6.516813     15.838078    9986.706258
std        6.749456      3.449624      8.859636   28840.649834
min     1969.000000      1.000000      1.000000       1.000000
25%     1974.000000      4.000000      8.000000    4388.000000
50%     1979.000000      7.000000     16.000000    4837.000000
75%     1984.000000     10.000000     23.000000    5300.000000
max     2008.000000     12.000000     31.000000  199622.000000

In [37]:
#fill NaNs with column mean in 'day' column
df['day'] = df['day'].fillna(df['day'].mean())

#view updated summary of null values
df.isnull().sum()

year      0
month     0
day       0
gender    0
births    0
dtype: int64

In [52]:
#check if months 1,3,5,7,8,10,12 have 31 days or less
days_31 = [1,3,5,7,8,10,12]
#check if month 2 has 28 or 29 days
#check if months 9,4,6 and 11 have days equal or less than 30
days_30 = [9,4,6,11]
df.loc[(df['month'] == 1) & (df['day'] > 31),'Days with error'] = 'Yes'
df.loc[(df['month'] == 3) & (df['day'] > 31),'Days with error'] = 'Yes'
df.loc[(df['month'] == 5) & (df['day'] > 31),'Days with error'] = 'Yes'
df.loc[(df['month'] == 8) & (df['day'] > 31),'Days with error'] = 'Yes'
df.loc[(df['month'] == 10) & (df['day'] > 31),'Days with error'] = 'Yes'
df.loc[(df['month'] == 12) & (df['day'] > 31),'Days with error'] = 'Yes'
df.loc[(df['month'] == 9) & (df['day'] > 30),'Days with error'] = 'Yes'
df.loc[(df['month'] == 4) & (df['day'] > 30),'Days with error'] = 'Yes'
df.loc[(df['month'] == 6) & (df['day'] > 30),'Days with error'] = 'Yes'
df.loc[(df['month'] == 11) & (df['day'] > 30),'Days with error'] = 'Yes'


In [53]:
df

year  month        day gender  births Days with error
0      1969      1   1.000000      F    4046             NaN
1      1969      1   1.000000      M    4440             NaN
2      1969      1   2.000000      F    4454             NaN
3      1969      1   2.000000      M    4548             NaN
4      1969      1   3.000000      F    4548             NaN
...     ...    ...        ...    ...     ...             ...
15542  2008     10  15.838078      M  183219             NaN
15543  2008     11  15.838078      F  158939             NaN
15544  2008     11  15.838078      M  165468             NaN
15545  2008     12  15.838078      F  173215             NaN
15546  2008     12  15.838078      M  181235             NaN

[15197 rows x 6 columns]

In [54]:
df.isnull().sum()

year                   0
month                  0
day                    0
gender                 0
births                 0
Days with error    15126
dtype: int64